solidity/contracts/bridge/library/FraudVerification.sol

```diff
@@ -0,0 +1,285 @@
+        bytes32 sighash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    );
+
+    event FraudChallengeDefended(
+        bytes walletPublicKey,
+        bytes32 sighash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    );
+
+    /// @notice Submits a wallet fraud challenge.
+    /// @param walletPublicKey The public key of the wallet in the uncompressed
+    ///                        and unprefixed format (64 bytes).
+    /// @param sighash Hash of the data (preimage) used for generating signature
+    ///                for the given input.
+    /// @param v Signature recovery value.
+    /// @param r Signature r value.
+    /// @param s Signature s value.
+    function submitFraudChallenge(
+        Data storage data,
+        bytes memory walletPublicKey,
+        bytes32 sighash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external {
+        // TODO: Bitcoin uses signatures in the DER format. From a signature in
+        // DER form values r and s can be easily extracted. Check how to obtain
+        // the v value. It can take values: 27, 28, 29, 30, 31, 32, 33, 34.
+        // https://bitcoin.stackexchange.com/questions/38351/ecdsa-v-r-s-what-is-v
+        require(
+            msg.value >= data.parameters.fraudChallengeDepositAmount,
+            "The amount of ETH deposited is too low"
+        );
+
+        bool verificationResult =
+            CheckBitcoinSigs.checkSig(walletPublicKey, sighash, v, r, s);
+        require(verificationResult, "Signature verification failure");
+
+        uint256 challengeKey =
+            uint256(
+                keccak256(abi.encodePacked(walletPublicKey, sighash, v, r, s))
+            );
+
+        FraudChallenge storage fraudChallenge =
+            data.fraudChallenges[challengeKey];
+        require(fraudChallenge.reportedAt == 0, "Fraud already challenged");
+
+        fraudChallenge.challenger = msg.sender;
+        fraudChallenge.ethDepositAmount = msg.value;
+        /* solhint-disable-next-line not-rely-on-time */
+        fraudChallenge.reportedAt = uint32(block.timestamp);
+        fraudChallenge.defended = false;
+
+        // TODO: Consider emitting the event with walletPublicKey in the
+        //       compressed format as it's how we identify wallets in the Bridge.
+        emit FraudChallengeSubmitted(walletPublicKey, sighash, v, r, s);
+    }
+
+    // TODO: description
+    function submitFraudChallengeResponse(
+        Data storage data,
+        bytes memory walletPublicKey,
+        bytes memory preimage,
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        bool witness,
+        mapping(uint256 => Bridge.DepositRequest) storage deposits,
+        mapping(uint256 => bool) storage spentMainUTXOs
+    ) external {
+        // TODO: Consider moving `deposits` and `spentMainUTXOs` to this library
+        // and renaming it `UTXO`. It would be responsible for UTXO-related
+        // logic, like UTXO frauds.
```

Comment:
As you realized, passing storage pointers to such critical mappings like `deposits` and `spentMainUTXOs` is dangerous as they can be modified unintentionally by the library code. On the other hand, copying them to memory may become very expensive at some point. I think we have two options:

1. Make `Fraud.sol` a very light library that does not modify any state but just prepares data for `Bridge.sol` that manages the state entirely. That means `submitFraudChallenge` should just validate the sighash+sig pair and return a `FraudChallenge` memory object to `Bridge` which does the rest. In that case `submitFraudChallengeResponse/defeatFraudChallenge` should just extract the `utxoKey` and return it to the `Bridge`. The `Bridge` should do all the validation of UTXO against its state, do the ETH bookkeeping and challenge lifecycle management.

2. Keep it as is and try to extract just the problematic parts. Here we can just split the `submitFraudChallengeResponse/defeatFraudChallenge` into two parts. First, `unwrapFraudChallenge` which will take the preimage, validate it against a challenge, and then extract the `utxoKey`. Second, the actual `defeatFraudChallenge` function will close the challenge and do the ETH accounting.

What do you think? Curious about @pdyraga opinion as well.